packages/admin/admin/src/common/buttons/feedback/FeedbackButton.tsx

```diff
@@ -0,0 +1,174 @@
+
+type OwnerState = { displayState: FeedbackButtonDisplayState };
+
+const Root = createComponentSlot(Button)<FeedbackButtonClassKey, OwnerState>({
+    componentName: "FeedbackButton",
+    slotName: "root",
+    classesResolver(ownerState) {
+        return [
+            ownerState.displayState === "idle" && "idle",
+            ownerState.displayState === "fail" && "fail",
+            ownerState.displayState === "success" && "success",
+            ownerState.displayState === "loading" && "loading",
+        ];
+    },
+})();
+
+const Tooltip = createComponentSlot(CometTooltip)<FeedbackButtonClassKey>({
+    componentName: "FeedbackButton",
+    slotName: "tooltip",
+})();
+
+export interface FeedbackButtonProps
+    extends ThemedComponentBaseProps<{
+            root: typeof Button;
+            tooltip: typeof CometTooltip;
+        }>,
+        ButtonProps {
+    loading?: boolean;
+    hasErrors?: boolean;
+    loadingIcon?: React.ReactNode;
+    startIcon?: React.ReactNode;
+    endIcon?: React.ReactNode;
+    tooltipSuccessMessage?: string;
+    tooltipErrorMessage?: string;
+}
+
+type FeedbackButtonDisplayState = "idle" | "loading" | "success" | "fail";
+
+export function FeedbackButton(inProps: FeedbackButtonProps) {
+    const {
+        loading,
+        hasErrors,
+        children,
+        variant = "contained",
+        color = "primary",
+        classes,
+        disabled,
+        loadingIcon = <ThreeDotSaving />,
+        startIcon,
+        endIcon,
+        tooltipSuccessMessage,
+        tooltipErrorMessage,
+        slotProps,
+
+        ...restProps
+    } = useThemeProps({
+        props: inProps,
+        name: "CometAdminFeedbackButton",
+    });
+
+    const [displayState, setDisplayState] = React.useState<FeedbackButtonDisplayState>("idle");
+
+    const ownerState: OwnerState = {
+        displayState: displayState,
+    };
+
+    const resolveIconForDisplayState = (displayState: FeedbackButtonDisplayState) => {
+        if (startIcon && displayState !== "loading") {
+            return startIcon;
+        } else if (endIcon && displayState !== "loading") {
+            return endIcon;
+        }
+        if (displayState === "loading") {
+            return loadingIcon;
+        }
+    };
+
+    const resolveTooltipForDisplayState = (displayState: FeedbackButtonDisplayState) => {
+        if (displayState === "success") {
+            return "success";
+        } else if (displayState === "fail") {
+            return "error";
+        } else {
+            return "neutral";
+        }
+    };
+
+    React.useEffect(() => {
+        let timeoutId: number | undefined;
+        let timeoutDuration: number | undefined;
+        let newDisplayState: FeedbackButtonDisplayState;
+
+        if (displayState === "idle" && loading) {
+            setDisplayState("loading");
+        } else if (displayState === "loading" && hasErrors) {
+            timeoutDuration = 500;
+            newDisplayState = "loading";
```

Comment:
What's the idea behind this case?

Reply:
Also in this case, I used the cases as they are used in the SaveButton. What is your suggestion to improve the cases?